src/coreclr/jit/emitxarch.cpp

```diff
@@ -144,6 +144,97 @@ bool emitter::IsDstSrcSrcAVXInstruction(instruction ins)
     return ((CodeGenInterface::instInfo[ins] & INS_Flags_IsDstSrcSrcAVXInstruction) != 0) && IsAVXInstruction(ins);
 }
 
+//------------------------------------------------------------------------
+// DoesWritesZeroFlag: check if the instruction write the
+//     ZF flag.
+//
+// Arguments:
+//    ins - instruction to test
+//
+// Return Value:
+//    true if instruction writes the ZF flag, false otherwise.
+//
+bool emitter::DoesWritesZeroFlag(instruction ins)
+{
+    return (CodeGenInterface::instInfo[ins] & INS_FLAGS_WritesZF) != 0;
+}
+
+//------------------------------------------------------------------------
+// DoesResetsOverflowAndCarryFlags: check if the instruction resets the
+//     OF and CF flag.
+//
+// Arguments:
+//    ins - instruction to test
+//
+// Return Value:
+//    true if instruction resets the OF and CF flag, false otherwise.
+//
+bool emitter::DoesResetsOverflowAndCarryFlags(instruction ins)
+{
+    return (CodeGenInterface::instInfo[ins] & INS_FLAGS_Resets_CF_OF_Flags) != 0;
+}
+
+//------------------------------------------------------------------------
+// IsFlagsAlwaysModified: check if the instruction always modifies the flags.
```

Comment:
```suggestion
// AreFlagsAlwaysModified: check if the instruction always modifies the flags.
```
I am a bit confused on what this is supposed to check. Does it mean: return true if there is at least one flag bit that is always modified by this instruction?

Seems odd we would ask that question instead of asking whether some specific set of flags is always modified.

Reply:
For shift instructions, if a shift-amount is `0`, it keeps the flags unmodified. For them, we shouldn't rely on last instruction, but the flags might have set by one of the previous instructions. As such, this method returns `false` saying that these instructions (under certain circumstances) guaranteed to not modify flags and the caller shouldn't rely on last instruction to eliminate `test`. For remaining instructions, they modify *some* flags and can eliminate `test` safely. I will add more explicit comment.

See my comment https://github.com/dotnet/runtime/pull/53214#discussion_r639097022 when I was not handling this case.